File: mobile/lib/features/payments/data/models/payment.dart
enum PaymentProvider {
  khalti,
  esewa,
  stripe,
  paypal;

  static PaymentProvider fromString(String v) {
    return PaymentProvider.values.firstWhere(
      (e) => e.name == v.toLowerCase(),
      orElse: () => PaymentProvider.khalti,
    );
  }

  String get displayName {
    switch (this) {
      case PaymentProvider.khalti:
        return 'Khalti';
      case PaymentProvider.esewa:
        return 'eSewa';
      case PaymentProvider.stripe:
        return 'Stripe';
      case PaymentProvider.paypal:
        return 'PayPal';
    }
  }
}

enum PaymentStatus {
  pending,
  initiated,
  completed,
  failed,
  refunded,
  cancelled;

  static PaymentStatus fromString(String v) {
    return PaymentStatus.values.firstWhere(
      (e) => e.name == v.toLowerCase(),
      orElse: () => PaymentStatus.pending,
    );
  }
}

class InitiatePaymentRequest {
  final PaymentProvider provider;
  final int amount;
  final String purchaseOrderId;
  final String purchaseOrderName;
  final String returnUrl;
  final String websiteUrl;
  final String? customerName;
  final String? customerEmail;
  final String? customerPhone;

  const InitiatePaymentRequest({
    required this.provider,
    required this.amount,
    required this.purchaseOrderId,
    required this.purchaseOrderName,
    required this.returnUrl,
    required this.websiteUrl,
    this.customerName,
    this.customerEmail,
    this.customerPhone,
  });

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      'provider': provider.name,
      'amount': amount,
      'purchase_order_id': purchaseOrderId,
      'purchase_order_name': purchaseOrderName,
      'return_url': returnUrl,
      'website_url': websiteUrl,
    };
    if (customerName != null) map['customer_name'] = customerName;
    if (customerEmail != null) map['customer_email'] = customerEmail;
    if (customerPhone != null) map['customer_phone'] = customerPhone;
    return map;
  }
}

class InitiatePaymentResponse {
  final int transactionId;
  final PaymentProvider provider;
  final PaymentStatus status;
  final String? paymentUrl;
  final String? providerPidx;
  final Map<String, dynamic>? extra;

  const InitiatePaymentResponse({
    required this.transactionId,
    required this.provider,
    required this.status,
    this.paymentUrl,
    this.providerPidx,
    this.extra,
  });

  factory InitiatePaymentResponse.fromJson(Map<String, dynamic> json) {
    return InitiatePaymentResponse(
      transactionId: json['transaction_id'] as int,
      provider: PaymentProvider.fromString(json['provider'] as String? ?? 'khalti'),
      status: PaymentStatus.fromString(json['status'] as String? ?? 'pending'),
      paymentUrl: json['payment_url'] as String?,
      providerPidx: json['provider_pidx'] as String?,
      extra: json['extra'] as Map<String, dynamic>?,
    );
  }
}

class VerifyPaymentRequest {
  final PaymentProvider provider;
  final String? pidx;
  final String? oid;
  final String? refId;
  final String? data;
  final int? transactionId;

  const VerifyPaymentRequest({
    required this.provider,
    this.pidx,
    this.oid,
    this.refId,
    this.data,
    this.transactionId,
  });

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{'provider': provider.name};
    if (pidx != null) map['pidx'] = pidx;
    if (oid != null) map['oid'] = oid;
    if (refId != null) map['refId'] = refId;
    if (data != null) map['data'] = data;
    if (transactionId != null) map['transaction_id'] = transactionId;
    return map;
  }
}

class VerifyPaymentResponse {
  final int transactionId;
  final PaymentProvider provider;
  final PaymentStatus status;
  final int? amount;
  final String? providerTransactionId;
  final Map<String, dynamic>? extra;

  const VerifyPaymentResponse({
    required this.transactionId,
    required this.provider,
    required this.status,
    this.amount,
    this.providerTransactionId,
    this.extra,
  });

  factory VerifyPaymentResponse.fromJson(Map<String, dynamic> json) {
    return VerifyPaymentResponse(
      transactionId: json['transaction_id'] as int,
      provider: PaymentProvider.fromString(json['provider'] as String? ?? 'khalti'),
      status: PaymentStatus.fromString(json['status'] as String? ?? 'pending'),
      amount: json['amount'] as int?,
      providerTransactionId: json['provider_transaction_id'] as String?,
      extra: json['extra'] as Map<String, dynamic>?,
    );
  }
}

class PaymentTransaction {
  final int id;
  final PaymentProvider provider;
  final PaymentStatus status;
  final int amount;
  final String currency;
  final String purchaseOrderId;
  final String purchaseOrderName;
  final String? providerTransactionId;
  final String? providerPidx;
  final String returnUrl;
  final String websiteUrl;
  final String? failureReason;

  const PaymentTransaction({
    required this.id,
    required this.provider,
    required this.status,
    required this.amount,
    required this.currency,
    required this.purchaseOrderId,
    required this.purchaseOrderName,
    this.providerTransactionId,
    this.providerPidx,
    required this.returnUrl,
    required this.websiteUrl,
    this.failureReason,
  });

  factory PaymentTransaction.fromJson(Map<String, dynamic> json) {
    return PaymentTransaction(
      id: json['id'] as int,
      provider: PaymentProvider.fromString(json['provider'] as String? ?? 'khalti'),
      status: PaymentStatus.fromString(json['status'] as String? ?? 'pending'),
      amount: json['amount'] as int? ?? 0,
      currency: json['currency'] as String? ?? 'NPR',
      purchaseOrderId: json['purchase_order_id'] as String? ?? '',
      purchaseOrderName: json['purchase_order_name'] as String? ?? '',
      providerTransactionId: json['provider_transaction_id'] as String?,
      providerPidx: json['provider_pidx'] as String?,
      returnUrl: json['return_url'] as String? ?? '',
      websiteUrl: json['website_url'] as String? ?? '',
      failureReason: json['failure_reason'] as String?,
    );
  }
}

File: backend/src/apps/finance/schemas/payment.py
"""
Payment schemas — request/response validation for the finance API.
Includes generic schemas usable across all providers, plus provider-specific
schemas for Khalti and eSewa.
"""
from typing import Any, Optional
from pydantic import BaseModel, field_serializer, field_validator

from src.apps.finance.models.payment import PaymentProvider, PaymentStatus
from src.apps.iam.utils.hashid import decode_id, encode_id


# ---------------------------------------------------------------------------
# Generic / provider-agnostic schemas
# ---------------------------------------------------------------------------

class InitiatePaymentRequest(BaseModel):
    """Request body to initiate a payment regardless of provider."""
    provider: PaymentProvider
    amount: int  # smallest currency unit (paisa for NPR)
    purchase_order_id: str
    purchase_order_name: str
    return_url: str
    website_url: str = ""
    customer_name: Optional[str] = None
    customer_email: Optional[str] = None
    customer_phone: Optional[str] = None

    @field_validator("amount")
    @classmethod
    def amount_must_be_positive(cls, v: int) -> int:
        if v <= 0:
            raise ValueError("amount must be a positive integer (in smallest currency unit)")
        return v


class InitiatePaymentResponse(BaseModel):
    """Response after successfully initiating a payment."""
    transaction_id: int          # our internal DB id
    provider: PaymentProvider
    status: PaymentStatus
    payment_url: Optional[str] = None   # URL to redirect the user to
    provider_pidx: Optional[str] = None # Khalti pidx / eSewa ref
    extra: Optional[dict[str, Any]] = None  # provider-specific extras

    @field_serializer("transaction_id")
    def serialize_transaction_id(self, value: int) -> str:
        return encode_id(value)


class VerifyPaymentRequest(BaseModel):
    """Request body to verify / confirm a payment callback."""
    provider: PaymentProvider
    pidx: Optional[str] = None          # Khalti uses pidx
    oid: Optional[str] = None           # eSewa uses oid (our purchase_order_id)
    refId: Optional[str] = None         # eSewa legacy refId
    # eSewa v2 passes a base64-encoded `data` query param from callback
    data: Optional[str] = None
    transaction_id: Optional[int] = None  # our internal transaction id

    @field_validator("transaction_id", mode="before")
    @classmethod
    def decode_transaction_id(cls, value: int | str | None) -> int | None:
        if isinstance(value, str):
            decoded = decode_id(value)
            if decoded is None:
                raise ValueError("Invalid transaction_id")
            return decoded
        return value


class VerifyPaymentResponse(BaseModel):
    """Normalised verification response from any provider."""
    transaction_id: int
    provider: PaymentProvider
    status: PaymentStatus
    amount: Optional[int] = None
    provider_transaction_id: Optional[str] = None
    extra: Optional[dict[str, Any]] = None

    @field_serializer("transaction_id")
    def serialize_transaction_id(self, value: int) -> str:
        return encode_id(value)


class PaymentTransactionRead(BaseModel):
    """Read schema for a stored PaymentTransaction (used in GET endpoints)."""
    id: int
    provider: PaymentProvider
    status: PaymentStatus
    amount: int
    currency: str
    purchase_order_id: str
    purchase_order_name: str
    provider_transaction_id: Optional[str]
    provider_pidx: Optional[str]
    return_url: str
    website_url: str
    failure_reason: Optional[str]

    model_config = {"from_attributes": True}

    @field_serializer("id")
    def serialize_id(self, value: int) -> str:
        return encode_id(value)


# ---------------------------------------------------------------------------
# Khalti-specific schemas (v2 API)
# ---------------------------------------------------------------------------

class KhaltiInitiateRequest(BaseModel):
    """Payload sent to Khalti /epayment/initiate/ endpoint."""
    return_url: str
    website_url: str
    amount: int
    purchase_order_id: str
    purchase_order_name: str
    customer_info: Optional[dict[str, str]] = None


class KhaltiInitiateResponse(BaseModel):
    """Response from Khalti /epayment/initiate/."""
    pidx: str
    payment_url: str
    expires_at: Optional[str] = None
    expires_in: Optional[int] = None


class KhaltiLookupRequest(BaseModel):
    """Payload sent to Khalti /epayment/lookup/."""
    pidx: str


class KhaltiLookupResponse(BaseModel):
    """Response from Khalti /epayment/lookup/."""
    pidx: str
    total_amount: int
    status: str          # Completed | Pending | Expired | User canceled
    transaction_id: Optional[str] = None
    fee: Optional[int] = None
    refunded: bool = False


# ---------------------------------------------------------------------------
# eSewa-specific schemas (v2 API)
# ---------------------------------------------------------------------------

class EsewaInitiateData(BaseModel):
    """
    Fields required to build the eSewa v2 form POST.
    The merchant must compute the HMAC-SHA256 signature before submitting.
    """
    amount: int                  # in paisa  (e.g. 100 = 1 NPR? No, eSewa uses rupees directly)
    tax_amount: int = 0
    total_amount: int
    transaction_uuid: str        # our unique order id
    product_code: str            # merchant code e.g. EPAYTEST
    product_service_charge: int = 0
    product_delivery_charge: int = 0
    success_url: str
    failure_url: str
    signed_field_names: str = "total_amount,transaction_uuid,product_code"
    signature: str               # HMAC-SHA256 of signed_field_names values


class EsewaCallbackData(BaseModel):
    """
    Decoded payload from eSewa's base64-encoded callback `data` param.
    eSewa sends this to success_url?data=<base64>
    """
    transaction_code: Optional[str] = None
    status: Optional[str] = None           # COMPLETE / PENDING / FULL_REFUND
    total_amount: Optional[str] = None
    transaction_uuid: Optional[str] = None
    product_code: Optional[str] = None
    signed_field_names: Optional[str] = None
    signature: Optional[str] = None
